source/fab/language/fortran.py

```diff
@@ -394,17 +398,78 @@ def run(self) -> List[Path]:
                                       'found': end_name}
                         raise TaskException(
                             end_message.format(**end_values))
-        return []
 
 
-class FortranPreProcessor(Command):
+class FortranPreProcessor(SingleFileCommand):
 
     @property
     def as_list(self) -> List[str]:
         base_command = ['cpp', '-traditional-cpp', '-P']
-        file_args = [str(self._filename), str(self.output_filename)]
+        file_args = [str(self._filename), str(self.output[0])]
         return base_command + self._flags + file_args
 
     @property
-    def output_filename(self) -> Path:
-        return self._workspace / self._filename.with_suffix('.f90').name
+    def output(self) -> List[Path]:
+        return [self._workspace /
+                self._filename.with_suffix('.f90').name]
+
+
+class FortranCompiler(Command):
+
+    def __init__(self,
+                 filename: Path,
+                 workspace: Path,
+                 flags: List[str],
+                 prerequisites: List[Path]):
+        super().__init__(workspace, flags)
+        self._filename = filename
+        self._prerequisites = prerequisites
+
+    @property
+    def as_list(self) -> List[str]:
+        base_command = ['gfortran',
+                        '-c',
+                        '-J' + str(self._workspace),
+                        ]
+        file_args = [str(self._filename),
+                     '-o',
+                     str(self.output[0]),
+                     ]
+        return base_command + self._flags + file_args
+
+    @property
+    def input(self) -> List[Path]:
+        return self._prerequisites + [self._filename]
+
+    @property
+    def output(self) -> List[Path]:
+        object_file = (
+            self._workspace / self._filename.with_suffix('.o').name)
+        return [object_file]
```

Comment:
It doesn't... but I take your point - for it to support module files it would need another input to the constructor to tell it what module filenames it is meant to be producing... I can do that

Reply:
I'm not sure how easy this right now, actually...  You would need to do something like this (in `application.py`):

1. Just prior to initialising the Compiler class, you'd have to query the database for the file you are about to compiler to find out all of the programming units it contains

2. For any of those units which are `module`s you could then generate `.mod` filenames and pass them to the constructor of the Compiler class

Unfortunately we don't currently know which units in the database are modules... I could add another TODO here which points out that it's something we want to do though?

Reply:
(now that I type out the above I realise there is already an incorrect assumption in what I've done because of this - assuming the Analyser can return non-module'd subroutines amongst the dependencies the current code will incorrectly assume it should expect a `.mod` file)

Reply:
Add the TODO please. We sort of knew we would need to add program unit type information to the database.

Reply:
Okay - I'll add 2 actually, one to point out the flaw in what's already there for the mod prerequisites, and one to point out that we want to do it for mod products also

Reply:
I don't see the TODO comments.

Reply:
Ah!  Okay yes, I put them where the constructor is called in `application.py` rather than here - I can add one here as well if you want?

Reply:
They have been pointed out to me at lines 124 and 132.